zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs

```diff
@@ -387,46 +496,294 @@ fn check_orchard_subtrees(db: &ZebraDb) -> bool {
         );
+
+        // Get the missing notes needed to complete the subtree.
+        //
+        // TODO: consider just reading the block's transactions from the database file,
+        //       because we don't use the block header data at all.
+        let block = read_db
+            .block(end_height.into())
+            .expect("height with note commitment tree should have block");
+        let sapling_note_commitments = block
+            .sapling_note_commitments()
+            .take(prev_remaining_notes)
+            .cloned()
+            .collect();
+
+        // This takes less than 1 second per tree, so we don't need to make it cancellable.
+        let (sapling_nct, subtree) = NoteCommitmentTrees::update_sapling_note_commitment_tree(
+            prev_tree,
+            sapling_note_commitments,
+        )
+        .expect("finalized notes should append successfully");
+
+        let (index, node) = subtree.unwrap_or_else(|| {
+            panic!(
+                "already checked that the block completed a subtree:\n\
+                 updated subtree:\n\
+                 index: {:?}\n\
+                 position: {:?}\n\
+                 remaining notes: {}\n\
+                 original previous subtree:\n\
+                 height: {prev_end_height:?}\n\
+                 index: {prev_index}\n\
+                 position: {prev_position}\n\
+                 remaining: {prev_remaining_notes}\n\
+                 original current subtree:\n\
+                 height: {end_height:?}\n\
+                 index: {current_index}\n\
+                 position: {current_position}\n\
+                 remaining: {current_remaining_notes}",
+                sapling_nct.subtree_index(),
+                sapling_nct.position(),
+                sapling_nct.remaining_subtree_leaf_nodes(),
+            )
+        });
+
+        NoteCommitmentSubtree::new(index, end_height, node)
     }
+}
 
-    is_valid
+/// Calculates a note commitment subtree for Orchard, reading blocks from `read_db` if needed.
+///
+/// The subtree must be completed by a note commitment in the block at `end_height`.
+/// `tree` is the tree for that block, and `prev_tree` is the tree for the previous block.
+///
+/// `prev_tree` is only used to rebuild the subtree if it was completed without using the last
+/// note commitment in the block at `end_height`.
+///
+/// # Panics
+///
+/// If a subtree is not completed by a note commitment in the block at `end_height`.
+#[must_use = "subtree should be written to the database after it is calculated"]
+#[instrument(skip(read_db, prev_tree, tree))]
+fn calculate_orchard_subtree(
+    read_db: &ZebraDb,
+    prev_end_height: Height,
+    prev_tree: Arc<orchard::tree::NoteCommitmentTree>,
+    end_height: Height,
+    tree: Arc<orchard::tree::NoteCommitmentTree>,
+) -> NoteCommitmentSubtree<orchard::tree::Node> {
+    // If a subtree is completed by a note commentment in the block at `end_height`,
```

Comment:
###### Optional Suggestion

```suggestion
    // If a subtree is completed by a note commitment in the block at `end_height`,
```

Reply:
This will get fixed when I copy-paste-replace the sapling comments:
https://github.com/ZcashFoundation/zebra/pull/7566#discussion_r1329046372